pytools/graph.py

```diff
@@ -240,6 +241,42 @@ def __init__(self, node: NodeT) -> None:
         self.node = node
 
 
+def find_cycles(graph: GraphT) -> List[List[NodeT]]:
+    """
+    Find all cycles in *graph* using DFS.
+
+    :returns: A :class:`list` in which each element represents another :class:`list`
+        of nodes that form a cycle.
+    """
+    def dfs(node: NodeT, path: List[NodeT]) -> List[NodeT]:
+        # Cycle detected
+        if visited[node] == 1:
+            return path
+
+        # Visit this node, explore its children
+        visited[node] = 1
+        path.append(node)
+        for child in graph[node]:
+            if visited[child] != 2 and dfs(child, path):
+                return path
+
+        # Done visiting node
+        visited[node] = 2
+        return []
+
+    visited = {node: 0 for node in graph.keys()}
+
+    res = []
+
+    for node in graph:
+        if not visited[node]:
+            cycle = dfs(node, [])
+            if cycle:
+                res.append(cycle)
```

Comment:
Finding all cycles is probably too much... all the use cases I can think of will only want one. I could see this as an option, maybe.

Reply:
I made it an optional parameter in 2ff7def. My thinking was that for the use case I had in mind, checking the validity of our DAG in pytato, having all cycles instead of just one may be beneficial.